### PR TITLE
feat: improve resolver interface

### PR DESCRIPTION
**Pull request by kitsonk**
This goes with a single enum response, for a resolver:

```rust
pub enum ResolveResponse {
  Amd(ModuleSpecifier),
  CommonJs(ModuleSpecifier),
  Err(Error),
  Esm(ModuleSpecifier),
  Script(ModuleSpecifier),
  Specifier(ModuleSpecifier),
  SystemJs(ModuleSpecifier),
  Umd(ModuleSpecifier),
}
```

### CODE REVIEW

File: src/source.rs [hunk 1]
```diff
@@ -89,10 +88,40 @@ pub trait Locker: fmt::Debug {
...
-#[derive(Debug, Clone, PartialEq)]
-pub struct ResolveResult {
-  pub specifier: ModuleSpecifier,
-  pub kind: ModuleKind,
+#[derive(Debug)]
+pub enum ResolveResponse {
+  Amd(ModuleSpecifier),
+  CommonJs(ModuleSpecifier),
+  Err(Error),
+  Esm(ModuleSpecifier),
```
Comment (lucacasonato) [inline src/source.rs:96]: Why is `Esm` separate from `Specifier`?

Comment (kitsonk) [reply src/source.rs:96]: `Specifier` indicates the resolver isn't really doing anything except resolving the specifier, where as `Esm` is asserting it is specifically an ESM module.  Current behaviour is that `Specifier` always gets assigned `ModuleKind::Esm` but we may want to change that in the future, or potentially perform further analysis on the module to determine the module kind.

Comment (lucacasonato) [reply src/source.rs:96]: Ok, this still confuses me, but let's do it.

Comment (bartlomieju) [reply src/source.rs:96]: @kitsonk could you add this info as doc comments?

**Review by bartlomieju — APPROVED** [review]
LGTM

**Review by dsherret — APPROVED** [review]
LGTM too

**Pull request merged**